Add auto configuration of account based on network resources.

dcaliste opened the pull request:
Complement the current auto configuration from
    settings with a network based approach inspired
    by the one used by Thunderbird.
    
    It first check if the domain serving the emails
    is exposing its configuration through an
    autoconfig XML file, otherwise it fallbacks to
    a database used by Thunderbird, maintained on
    Github. For associated domains, it finally
    relies on the autoconfig service hosted by
    Thunderbird.

@pvuorela, this is a proposition to have a better coverage of existing email servers than what the files in `configuration/` can provide. I noted it as a draft to open a discussion, but it's actually usable (but not tested on device yet, just the unit testing). The points I would like to raise are following:
1. I'm proposing an hybrid solution, first based on the existing QSettings files provided here and complementing it by on-line resources,
2. should we simply rely on the QSettings and update them with a script based on these online resources ?
3. should we completely drop the QSettings and only rely on the online resources from device ?

Point 2. has the drawback to require to update it from time to time and miss the advertised config served by provider themselves when they are using the autoconfig mechanism. But it has the advantage to be totally offline on device.

Point 3. has the advantage to completely externalise the autoconfig stuff, but it requires to be online when creating an account from the UI (which may be the case most of the time, just to get the emails for the account).

Point 1. is a middle ground, having the disavantage to maintain a local copy that may not be in sync with the online resources and thus keeping the burden to maintain it.

The changes in the code I'm proposing can easily allow point 3. without much modifications.

I'm also proposing this, because my two providers (free.fr and mailbox.org) are not in the current QSettings. While I can add them easily there, I was trying to get something wider and more future proof. What do you think ?

PS : it is currently based on #35 for ease of testing from my side. It can be rebased on master easily or wait for #35 to be polished and accepted. So just the last commit is relevant here.

pvuorela commented:
Nice! The less we need to keep email service specific data the better.

Out of the options, like Thuderbird wiki documents their usage, I think we could keep both. There could be reasons to need config which isn't available online. Any manual sync of the resources I'd rather skip as that's destined to be outdated at points and manual work is bad.

Network access needs consideration, but I think if someone configures an email account without filling in anything more than address and password, it should be quite ok accessing the service itself, and maybe that's even opting in enough for asking thunderbird services.

Another question is which should have higher priority if both exist, the network config or the local config. The service itself I could trust the most for being accurate.

With the autoconfig there probably is something we could then remove from the local configuration data.

pvuorela commented:
This probably has less resistance of merging than the alias support so might be a good idea to do the PR without anything else.

Does this have some concerns for 'draft' status or is it just about the discussion on which path we should go with the auto configuration mechanisms?

dcaliste commented:
> So for now I'd just make the network data have higher priority and later if we need to support overriding we can implement the config option for priority order.

I've moved the local setting as a fallback of the network settings (no network or nothing found). I've adjusted the test accordingly.

pvuorela commented on tests/tests.pro:
```diff
@@ -2,8 +2,8 @@ TEMPLATE = subdirs
-    tst_folderlistmodel
-    
+    tst_folderlistmodel \
+    tst_autoconfig
```
Now that there is some detection for network, maybe you could add the unit test to the tests.xml file? I usually just run that with the testrunner on the device to ensure the tests still pass.

dcaliste replied:
Sorry, I forgot about this addition yesterday. It's corrected now.

dcaliste commented:
> if accessing other services would be considered suspicious enough, I'd then just remove the urls from the list.

Ok, so let's keep it as it is for the moment. Thanks for the review and comments.

pvuorela commented on src/emailautoconfig.cpp:
```diff
@@ -0,0 +1,369 @@
+                                 QStringLiteral("authentication")));
+}
+
+EmailAutoConfig::AuthList EmailAutoConfig::popAuthentication(QMailTransport::EncryptType type) const
```
Hmh, these two seem only used by the unit test thouh smtp has usage in emailaccount. Maybe good enough but just noting.

dcaliste replied:
Well, I agree that EmailAccount is not using this information at the moment, and may never actually use it, relying on autodetection by CAPABILITIES in the protocols (pop or imap). But I prefer to be complete in this EmailAutoConfig class, providing accessors for every bit of information.

pvuorela reviewed (approved):
Alright, think this is good to go. Thanks!
